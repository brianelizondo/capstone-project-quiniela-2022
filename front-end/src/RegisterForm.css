.RegisterForm-form-card{
    padding: 30px 30px;
    background-color: rgba(0, 0, 0, .35);
    border-radius: 10px !important;
}

.RegisterForm-form-container small{
    color: lightgray !important;
    font-size: 12px;
}
.RegisterForm-form-container button{
    font-size: 20px;
    font-weight: 700;
    text-transform: uppercase;
}