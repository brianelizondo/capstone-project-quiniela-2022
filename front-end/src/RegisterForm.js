import React, { useState } from 'react';
import { useHistory } from "react-router-dom";
import { Form, FloatingLabel, Button, Spinner, Col, Row } from "react-bootstrap";
import { useFormik } from 'formik';
import Loading from './Loading';


function RegisterForm({ userRegister, checkUsernameEmail }){
    // history, loading and state for submit button
    const history = useHistory();
    const [btnSubmitLoading, setBtnSubmitLoading] = useState(false);
    const [currentUsernameEmail, setCurrentUsernameEmail] = useState({ username: "", email: "" });

    const validate = async values => {
        const errors = {};
        if(!values.firstName) {
            errors.firstName = 'Required';
        }else if(!/^([a-zA-Z ])*$/.test(values.firstName)){
            errors.firstName = "Only letters are allowed";
        }else if(values.firstName.length > 50) {
            errors.firstName = 'Must be 50 characters or less';
        }

        if(!values.lastName) {
            errors.lastName = 'Required';
        }else if(!/^([a-zA-Z ])*$/.test(values.lastName)){
            errors.lastName = "Only letters are allowed";
        }else if(values.lastName.length > 50) {
            errors.lastName = 'Must be 50 characters or less';
        }
        
        if(!values.email) {
            errors.email = 'Required';
        }else if(!/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(values.email)) {
            errors.email = 'Invalid email address';
        }else if(values.email.length > 50) {
            errors.email = 'Must be 50 characters or less';
        }else if(values.email !== "" && currentUsernameEmail.email !== values.email){
            setCurrentUsernameEmail({...currentUsernameEmail, email: values.email});
            if(await checkUsernameEmail({ username: false, email: values.email })){
                errors.email = 'The email already exist';
            }
        }

        if(!values.username) {
            errors.username = 'Required';
        }else if(!/^([a-zA-Z0-9])*$/.test(values.username)){
            errors.password = "Only letters and numbers are allowed";
        }else if(values.username.length < 8) {
            errors.username = 'Must be 8 characters or more';
        }else if(values.username.length > 20) {
            errors.username = 'Must be 20 characters or less';
        }else if(values.username !== "" && currentUsernameEmail.username !== values.username){
            setCurrentUsernameEmail({...currentUsernameEmail, username: values.username});
            if(await checkUsernameEmail({ username: values.username, email: false })){
                errors.username = 'The username already exist';
            }
        }

        if(!values.password) {
            errors.password = 'Required';
        }else if(!/^([a-zA-Z0-9])*$/.test(values.password)){
            errors.password = "Only letters and numbers are allowed";
        }else if(values.password.length < 8) {
            errors.password = 'Must be 8 characters or more';
        }else if(values.password.length > 20) {
            errors.password = 'Must be 20 characters or less';
        }
      
        return errors;
    };

    // config for user register form
    const formik = useFormik({
        initialValues: {
            firstName: "", 
            lastName: "", 
            email: "", 
            username: "", 
            password: ""
        },
        validate,
        onSubmit: async (values) => {
            setBtnSubmitLoading(true);
            try{
                userRegister(values);
                history.push('/login');
            }catch (err){
                history.push('404');
            }
        }
    });

    return (
        <div className="RegisterForm col-md-8 offset-md-2">
            <h1>Register</h1>
            <p>Register to Quinielas World Cup 2022</p>
            
            <div className="col-md-8 offset-md-2">
                <Form onSubmit={formik.handleSubmit}>
                    <Row>
                        <Col>
                            <FloatingLabel controlId="firstName" label="First Name" className="mb-3">
                                <Form.Control name="firstName" type="text" placeholder="First Name" 
                                    onChange={formik.handleChange} 
                                    onBlur={formik.handleBlur} 
                                    value={formik.values.firstName}
                                />
                                {formik.touched.firstName && formik.errors.firstName ? <div>{formik.errors.firstName}</div> : null}
                            </FloatingLabel>
                        </Col>
                        <Col>
                            <FloatingLabel controlId="lastName" label="Last Name" className="mb-3">
                                <Form.Control name="lastName" type="text" placeholder="First Name" 
                                    onChange={formik.handleChange} 
                                    onBlur={formik.handleBlur} 
                                    value={formik.values.lastName}
                                />
                                {formik.touched.lastName && formik.errors.lastName ? <div>{formik.errors.lastName}</div> : null}
                            </FloatingLabel>
                        </Col>
                    </Row>
                    <Row>
                        <Col>
                            <FloatingLabel controlId="email" label="Email" className="mb-3">
                                <Form.Control name="email" type="email" placeholder="Email" 
                                    onChange={formik.handleChange} 
                                    onBlur={formik.handleBlur} 
                                    value={formik.values.email}
                                />
                                {formik.touched.email && formik.errors.email ? <div>{formik.errors.email}</div> : null}
                            </FloatingLabel>
                        </Col>
                    </Row>
                    <Row>
                        <Col>
                            <FloatingLabel controlId="username" label="Username" className="mb-3">
                                <Form.Control name="username" type="text" placeholder="Username" 
                                    onChange={formik.handleChange} 
                                    onBlur={formik.handleBlur} 
                                    value={formik.values.username}
                                />
                                {formik.touched.username && formik.errors.username ? <div>{formik.errors.username}</div> : <Form.Text id="usernameHelpBlock" muted>Your username must be 8-20 characters long and contain letters and numbers</Form.Text>}
                            </FloatingLabel>
                        </Col>
                    </Row>
                    <Row>
                        <Col>
                            <FloatingLabel controlId="password" label="Password" className="mb-3">
                                <Form.Control name="password" type="password" placeholder="Password" 
                                    onChange={formik.handleChange} 
                                    onBlur={formik.handleBlur} 
                                    value={formik.values.password}
                                />
                                {formik.touched.password && formik.errors.password ? <div>{formik.errors.password}</div> : <Form.Text id="passwordHelpBlock" muted>Your password must be 8-20 characters long, contain letters and numbers, and must not contain spaces, special characters or emoji</Form.Text>}
                            </FloatingLabel>
                        </Col>
                    </Row>
                    <Row>
                        <Col>
                        <Button variant="primary" type="submit" size="sm" disabled={ btnSubmitLoading }>
                            { btnSubmitLoading ? <Spinner as="span" animation="border" size="sm" role="status" aria-hidden="true" /> : "Register!" }
                        </Button>
                        </Col>
                    </Row>
                </Form>
            </div>
        </div>
    );
}

export default RegisterForm;