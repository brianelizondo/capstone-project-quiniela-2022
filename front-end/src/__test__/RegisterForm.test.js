import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import { MemoryRouter } from "react-router-dom";
import userEvent from '@testing-library/user-event';

import RegisterForm from '../RegisterForm';

// Smoke Test
describe("Smoke Test", () => {
    it("renders 'RegisterForm' component without crashing", () => {
        const userRegister = jest.fn();
        const checkUsernameEmail = jest.fn();
        render(
            <MemoryRouter initialEntries={["/register"]}>
                <RegisterForm userRegister={userRegister} checkUsernameEmail={checkUsernameEmail} />
            </MemoryRouter>
        );
        const textFound = screen.getByText("Register to Quinielas World Cup 2022");
        expect(textFound).toBeInTheDocument();
    });
});

// Snapshot Tests
describe("Snapshot Test", () => {
    it("'RegisterForm' matches snapshot", () => {
        const userRegister = jest.fn();
        const checkUsernameEmail = jest.fn();
        const {asFragment} = render(
            <MemoryRouter initialEntries={["/register"]}>
                <RegisterForm userRegister={userRegister} checkUsernameEmail={checkUsernameEmail} />
            </MemoryRouter>
        );
        expect(asFragment()).toMatchSnapshot();
    });
});

describe("Additional Test", () => {
    it("should show a required field warning for each empty input field", async () => {
        const userRegister = jest.fn();
        const checkUsernameEmail = jest.fn();
        render(
            <MemoryRouter initialEntries={["/register"]}>
                <RegisterForm userRegister={userRegister} checkUsernameEmail={checkUsernameEmail} />
            </MemoryRouter>
        );
        const user = userEvent.setup();
    
        await user.click(screen.getByRole("button", {type: /Register!/i}));
        const warnings = await screen.findAllByText("Required");
        expect(warnings[0]).toBeVisible();
    });

    it("should show invalid field errors for each invalid input field", async () => {
        const userRegister = jest.fn();
        const checkUsernameEmail = jest.fn();
        render(
            <MemoryRouter initialEntries={["/register"]}>
                <RegisterForm userRegister={userRegister} checkUsernameEmail={checkUsernameEmail} />
            </MemoryRouter>
        );
        const user = userEvent.setup();

        await user.type(screen.getByLabelText(/First Name/i), '1234');
        await user.type(screen.getByLabelText(/Last Name/i), '1234');
        await user.click(screen.getByRole("button", {type: /Register!/i}));
        let warnings = await screen.findAllByText("Only letters are allowed");
        expect(warnings[0]).toBeVisible();

        await user.type(screen.getByLabelText(/Email/i), '1234');
        await user.click(screen.getByRole("button", {type: /Register!/i}));
        warnings = await screen.findAllByText("Invalid email address");
        expect(warnings[0]).toBeVisible();

        await user.type(screen.getByLabelText(/Username/i), '1234');
        await user.type(screen.getByLabelText(/Password/i), '1234');
        await user.click(screen.getByRole("button", {type: /Register!/i}));
        warnings = await screen.findAllByText("Must be 8 characters or more");
        expect(warnings[0]).toBeVisible();

        await user.type(screen.getByLabelText(/Username/i), '1234....');
        await user.type(screen.getByLabelText(/Password/i), '1234....');
        await user.click(screen.getByRole("button", {type: /Register!/i}));
        warnings = await screen.findAllByText("Only letters and numbers are allowed");
        expect(warnings[0]).toBeVisible();
    });
});