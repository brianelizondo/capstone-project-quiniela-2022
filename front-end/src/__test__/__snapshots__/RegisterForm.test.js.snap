// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Test 'RegisterForm' matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="RegisterForm col-md-8 offset-md-2"
  >
    <h1>
      Register
    </h1>
    <p>
      Register to Quinielas World Cup 2022
    </p>
    <div
      class="col-md-8 offset-md-2"
    >
      <form
        class=""
      >
        <div
          class="row"
        >
          <div
            class="col"
          >
            <div
              class="mb-3 form-floating"
            >
              <input
                class="form-control"
                id="firstName"
                name="firstName"
                placeholder="First Name"
                type="text"
                value=""
              />
              <div
                class="invalid-feedback"
              />
              <label
                for="firstName"
              >
                First Name
              </label>
            </div>
          </div>
          <div
            class="col"
          >
            <div
              class="mb-3 form-floating"
            >
              <input
                class="form-control"
                id="lastName"
                name="lastName"
                placeholder="First Name"
                type="text"
                value=""
              />
              <div
                class="invalid-feedback"
              />
              <label
                for="lastName"
              >
                Last Name
              </label>
            </div>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="col"
          >
            <div
              class="mb-3 form-floating"
            >
              <input
                class="form-control"
                id="email"
                name="email"
                placeholder="Email"
                type="email"
                value=""
              />
              <div
                class="invalid-feedback"
              />
              <label
                for="email"
              >
                Email
              </label>
            </div>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="col"
          >
            <div
              class="mb-3 form-floating"
            >
              <input
                class="form-control"
                id="username"
                name="username"
                placeholder="Username"
                type="text"
                value=""
              />
              <div
                class="invalid-feedback"
              />
              <small
                class="form-text text-muted"
                id="usernameHelpBlock"
              >
                Your username must be 8-20 characters long and contain letters and numbers
              </small>
              <label
                for="username"
              >
                Username
              </label>
            </div>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="col"
          >
            <div
              class="mb-3 form-floating"
            >
              <input
                class="form-control"
                id="password"
                name="password"
                placeholder="Password"
                type="password"
                value=""
              />
              <div
                class="invalid-feedback"
              />
              <small
                class="form-text text-muted"
                id="passwordHelpBlock"
              >
                Your password must be 8-20 characters long, contain letters and numbers, and must not contain spaces, special characters or emoji
              </small>
              <label
                for="password"
              >
                Password
              </label>
            </div>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="col"
          >
            <button
              class="btn btn-primary btn-sm"
              type="submit"
            >
              Register!
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
